src/cpp/src/scheduler.hpp

```diff
@@ -45,6 +52,23 @@ class Scheduler {
 
     Output schedule(std::vector<SequenceGroup::Ptr>& sequence_groups) {
         Output scheduler_output;
+        float eps = 1e-5;
+
+        if (!m_block_manager.block_allocator_initialized()) {
+            size_t prompt_sum_size = 0;
+            for (auto idx = 0; idx < sequence_groups.size(); idx++) {
+                prompt_sum_size += sequence_groups[idx]->get_prompt_len();
+            }
+            size_t initial_kv_cache_size = prompt_sum_size * m_kv_blocks_initial_multiplier;
+            m_block_manager.increase_kv_blocks_number(initial_kv_cache_size);
+            m_dynamic_memory_allocation = true;
+        }
+        else if (m_dynamic_memory_allocation && (m_block_manager.get_used_percentage() + eps) > m_precentage_threshold_for_cache_increase) {
+            size_t new_cache_size = (size_t)(m_block_manager.get_total_number_of_kv_blocks() * m_cache_growth_factor);
+            m_block_manager.increase_kv_blocks_number(new_cache_size);
+        }
+        OPENVINO_ASSERT(m_cache_manager != nullptr, "Cache manager needs to be set in the Scheduler constructor.");
+        m_cache_manager->allocate_cache_if_needed(m_block_manager.get_total_number_of_kv_blocks());
```

Comment:
maybe we can move this logic to `ContBatchingPipeline::step()` ? as it does not seem `Scheduler` responsibility to re-allocate KV cache.

I would create a method `ContBatchingPipeline::reallocate_kv_cache_if_needed` or similar and use it within `step()`.

`Scheduler` can expose `block_manager` if it's done yet.